// @tambo-ai/react/esm/v1/providers/tambo-v1-stream-context.js
"use client";
/**
 * Stream Context Provider for v1 API
 *
 * Manages streaming state using React Context and useReducer.
 * Provides state and dispatch to child components via separate contexts
 * following the split-context pattern for optimal re-render performance.
 */
// React is used implicitly for JSX transformation (jsx: "react" in tsconfig)
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import React, { createContext, useReducer, useContext, useMemo, useCallback, } from "react";
import { streamReducer, createInitialThreadState, } from "../utils/event-accumulator";
/**
 * Context for accessing stream state (read-only).
 * Separated from dispatch context to prevent unnecessary re-renders.
 */
const StreamStateContext = createContext(null);
/**
 * Context for dispatching events to the stream reducer.
 * Separated from state context to prevent unnecessary re-renders.
 */
const StreamDispatchContext = createContext(null);
/**
 * Context for thread management functions.
 * Separated from state to prevent unnecessary re-renders.
 */
const ThreadManagementContext = createContext(null);
/**
 * Creates initial stream state from props.
 * @param props - Provider props
 * @returns Initial stream state
 */
function createInitialState(props) {
    const { initialThread, threadId } = props;
    // Initialize with empty threadMap
    const threadMap = {};
    // If threadId is provided, initialize that thread
    if (threadId) {
        // Create initial thread state (immutably)
        const baseState = createInitialThreadState(threadId);
        // If initial thread data provided, merge it immutably
        const threadState = initialThread
            ? {
                ...baseState,
                thread: {
                    ...baseState.thread,
                    ...initialThread,
                    id: threadId, // Always use the provided threadId
                },
            }
            : baseState;
        threadMap[threadId] = threadState;
    }
    return {
        threadMap,
        currentThreadId: threadId ?? null,
    };
}
/**
 * Provider component for stream state management.
 *
 * Uses useReducer with streamReducer to accumulate AG-UI events into
 * thread state. Provides state, dispatch, and thread management via separate contexts.
 * @returns JSX element wrapping children with stream contexts
 * @example
 * ```tsx
 * <TamboV1StreamProvider threadId="thread_123">
 *   <ChatInterface />
 * </TamboV1StreamProvider>
 * ```
 */
export function TamboV1StreamProvider(props) {
    const { children } = props;
    const initialState = useMemo(() => createInitialState(props), 
    // Only recompute if threadId changes
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [props.threadId]);
    const [state, dispatch] = useReducer(streamReducer, initialState);
    // Thread management functions
    const initThread = useCallback((threadId, initialThread) => {
        dispatch({ type: "INIT_THREAD", threadId, initialThread });
    }, []);
    const switchThread = useCallback((threadId) => {
        dispatch({ type: "SET_CURRENT_THREAD", threadId });
    }, []);
    const startNewThread = useCallback(() => {
        const tempId = `temp_${crypto.randomUUID()}`;
        // Use atomic START_NEW_THREAD action to prevent race conditions
        // when multiple calls happen concurrently (e.g., double-click)
        dispatch({ type: "START_NEW_THREAD", threadId: tempId });
        return tempId;
    }, []);
    const threadManagement = useMemo(() => ({
        initThread,
        switchThread,
        startNewThread,
    }), [initThread, switchThread, startNewThread]);
    return (React.createElement(StreamStateContext.Provider, { value: state },
        React.createElement(StreamDispatchContext.Provider, { value: dispatch },
            React.createElement(ThreadManagementContext.Provider, { value: threadManagement }, children))));
}
/**
 * Hook to access stream state.
 *
 * Must be used within TamboV1StreamProvider.
 * @returns Current stream state
 * @throws {Error} if used outside TamboV1StreamProvider
 * @example
 * ```tsx
 * function ChatMessages() {
 *   const { thread, streaming } = useStreamState();
 *
 *   return (
 *     <div>
 *       {thread.messages.map(msg => <Message key={msg.id} message={msg} />)}
 *       {streaming.status === 'streaming' && <LoadingIndicator />}
 *     </div>
 *   );
 * }
 * ```
 */
export function useStreamState() {
    const context = useContext(StreamStateContext);
    if (!context) {
        throw new Error("useStreamState must be used within TamboV1StreamProvider");
    }
    return context;
}
/**
 * Hook to access stream dispatch function.
 *
 * Must be used within TamboV1StreamProvider.
 * @returns Dispatch function for sending events to reducer
 * @throws {Error} if used outside TamboV1StreamProvider
 * @example
 * ```tsx
 * function StreamHandler() {
 *   const dispatch = useStreamDispatch();
 *
 *   useEffect(() => {
 *     async function handleStream() {
 *       for await (const event of streamEvents) {
 *         dispatch({ type: 'EVENT', event });
 *       }
 *     }
 *     handleStream();
 *   }, [dispatch]);
 *
 *   return null;
 * }
 * ```
 */
export function useStreamDispatch() {
    const context = useContext(StreamDispatchContext);
    if (!context) {
        throw new Error("useStreamDispatch must be used within TamboV1StreamProvider");
    }
    return context;
}
/**
 * Hook to access thread management functions.
 *
 * Must be used within TamboV1StreamProvider.
 * @returns Thread management functions
 * @throws {Error} if used outside TamboV1StreamProvider
 * @example
 * ```tsx
 * function ThreadSwitcher() {
 *   const { switchThread, startNewThread } = useThreadManagement();
 *
 *   return (
 *     <div>
 *       <button onClick={() => switchThread('thread_123')}>
 *         Load Thread
 *       </button>
 *       <button onClick={startNewThread}>
 *         New Chat
 *       </button>
 *     </div>
 *   );
 * }
 * ```
 */
export function useThreadManagement() {
    const context = useContext(ThreadManagementContext);
    if (!context) {
        throw new Error("useThreadManagement must be used within TamboV1StreamProvider");
    }
    return context;
}
