// @tambo-ai/react/dist/v1/index.js
"use strict";
/**
 * `@tambo-ai/react/v1` - React SDK for Tambo v1 API
 *
 * Provides React hooks and providers for building AI-powered applications
 * using the v1 streaming API with AG-UI protocol.
 *
 * ## Quick Start
 *
 * ```tsx
 * import {
 *   TamboV1Provider,
 *   useTamboV1,
 *   useTamboV1SendMessage,
 * } from '@tambo-ai/react/v1';
 *
 * function App() {
 *   return (
 *     <TamboV1Provider
 *       apiKey={process.env.NEXT_PUBLIC_TAMBO_API_KEY!}
 *       components={[WeatherCard]}
 *       tools={[searchTool]}
 *     >
 *       <ChatInterface />
 *     </TamboV1Provider>
 *   );
 * }
 *
 * function ChatInterface() {
 *   const [threadId, setThreadId] = useState<string>();
 *   const { messages, isStreaming } = useTamboV1(threadId);
 *   const sendMessage = useTamboV1SendMessage(threadId);
 *
 *   const handleSend = async (text: string) => {
 *     const result = await sendMessage.mutateAsync({
 *       message: { role: 'user', content: [{ type: 'text', text }] },
 *     });
 *     if (!threadId) setThreadId(result.threadId);
 *   };
 *
 *   return (
 *     <div>
 *       {messages.map(msg => <Message key={msg.id} message={msg} />)}
 *       {isStreaming && <LoadingIndicator />}
 *       <MessageInput onSend={handleSend} />
 *     </div>
 *   );
 * }
 * ```
 *
 * ## Type Imports
 *
 * Types are imported directly from specific files:
 * - Thread state: `import type { TamboV1Thread } from '@tambo-ai/react/v1/types/thread'`
 * - Messages: `import type { TamboV1Message } from '@tambo-ai/react/v1/types/message'`
 * - Custom events: `import type { ComponentStartEvent } from '@tambo-ai/react/v1/types/event'`
 *
 * SDK types: `import type { ... } from '@tambo-ai/typescript-sdk'`
 * AG-UI events: `import { EventType, type BaseEvent } from '@ag-ui/core'`
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.handleEventStream = exports.ToolCallTracker = exports.executeAllPendingTools = exports.executeClientTool = exports.toAvailableTools = exports.toAvailableTool = exports.toAvailableComponents = exports.toAvailableComponent = exports.applyJsonPatch = exports.useTamboV1ThreadList = exports.useTamboV1Thread = exports.useTamboV1SendMessage = exports.useTamboV1Messages = exports.useTamboV1 = exports.TamboRegistryProvider = exports.useThreadManagement = exports.useStreamDispatch = exports.useStreamState = exports.TamboV1StreamProvider = exports.TamboV1Provider = void 0;
// =============================================================================
// Providers
// =============================================================================
var tambo_v1_provider_1 = require("./providers/tambo-v1-provider");
Object.defineProperty(exports, "TamboV1Provider", { enumerable: true, get: function () { return tambo_v1_provider_1.TamboV1Provider; } });
var tambo_v1_stream_context_1 = require("./providers/tambo-v1-stream-context");
Object.defineProperty(exports, "TamboV1StreamProvider", { enumerable: true, get: function () { return tambo_v1_stream_context_1.TamboV1StreamProvider; } });
Object.defineProperty(exports, "useStreamState", { enumerable: true, get: function () { return tambo_v1_stream_context_1.useStreamState; } });
Object.defineProperty(exports, "useStreamDispatch", { enumerable: true, get: function () { return tambo_v1_stream_context_1.useStreamDispatch; } });
Object.defineProperty(exports, "useThreadManagement", { enumerable: true, get: function () { return tambo_v1_stream_context_1.useThreadManagement; } });
// Re-export registry provider from beta SDK (works with v1)
var tambo_registry_provider_1 = require("../providers/tambo-registry-provider");
Object.defineProperty(exports, "TamboRegistryProvider", { enumerable: true, get: function () { return tambo_registry_provider_1.TamboRegistryProvider; } });
// =============================================================================
// Hooks
// =============================================================================
var use_tambo_v1_1 = require("./hooks/use-tambo-v1");
Object.defineProperty(exports, "useTamboV1", { enumerable: true, get: function () { return use_tambo_v1_1.useTamboV1; } });
var use_tambo_v1_messages_1 = require("./hooks/use-tambo-v1-messages");
Object.defineProperty(exports, "useTamboV1Messages", { enumerable: true, get: function () { return use_tambo_v1_messages_1.useTamboV1Messages; } });
var use_tambo_v1_send_message_1 = require("./hooks/use-tambo-v1-send-message");
Object.defineProperty(exports, "useTamboV1SendMessage", { enumerable: true, get: function () { return use_tambo_v1_send_message_1.useTamboV1SendMessage; } });
var use_tambo_v1_thread_1 = require("./hooks/use-tambo-v1-thread");
Object.defineProperty(exports, "useTamboV1Thread", { enumerable: true, get: function () { return use_tambo_v1_thread_1.useTamboV1Thread; } });
var use_tambo_v1_thread_list_1 = require("./hooks/use-tambo-v1-thread-list");
Object.defineProperty(exports, "useTamboV1ThreadList", { enumerable: true, get: function () { return use_tambo_v1_thread_list_1.useTamboV1ThreadList; } });
// =============================================================================
// Utilities
// =============================================================================
var json_patch_1 = require("./utils/json-patch");
Object.defineProperty(exports, "applyJsonPatch", { enumerable: true, get: function () { return json_patch_1.applyJsonPatch; } });
var registry_conversion_1 = require("./utils/registry-conversion");
Object.defineProperty(exports, "toAvailableComponent", { enumerable: true, get: function () { return registry_conversion_1.toAvailableComponent; } });
Object.defineProperty(exports, "toAvailableComponents", { enumerable: true, get: function () { return registry_conversion_1.toAvailableComponents; } });
Object.defineProperty(exports, "toAvailableTool", { enumerable: true, get: function () { return registry_conversion_1.toAvailableTool; } });
Object.defineProperty(exports, "toAvailableTools", { enumerable: true, get: function () { return registry_conversion_1.toAvailableTools; } });
var tool_executor_1 = require("./utils/tool-executor");
Object.defineProperty(exports, "executeClientTool", { enumerable: true, get: function () { return tool_executor_1.executeClientTool; } });
Object.defineProperty(exports, "executeAllPendingTools", { enumerable: true, get: function () { return tool_executor_1.executeAllPendingTools; } });
var tool_call_tracker_1 = require("./utils/tool-call-tracker");
Object.defineProperty(exports, "ToolCallTracker", { enumerable: true, get: function () { return tool_call_tracker_1.ToolCallTracker; } });
var stream_handler_1 = require("./utils/stream-handler");
Object.defineProperty(exports, "handleEventStream", { enumerable: true, get: function () { return stream_handler_1.handleEventStream; } });
