// @tambo-ai/react/dist/v1/providers/tambo-v1-stream-context.js
"use strict";
"use client";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.TamboV1StreamProvider = TamboV1StreamProvider;
exports.useStreamState = useStreamState;
exports.useStreamDispatch = useStreamDispatch;
exports.useThreadManagement = useThreadManagement;
/**
 * Stream Context Provider for v1 API
 *
 * Manages streaming state using React Context and useReducer.
 * Provides state and dispatch to child components via separate contexts
 * following the split-context pattern for optimal re-render performance.
 */
// React is used implicitly for JSX transformation (jsx: "react" in tsconfig)
// eslint-disable-next-line @typescript-eslint/no-unused-vars
const react_1 = __importStar(require("react"));
const event_accumulator_1 = require("../utils/event-accumulator");
/**
 * Context for accessing stream state (read-only).
 * Separated from dispatch context to prevent unnecessary re-renders.
 */
const StreamStateContext = (0, react_1.createContext)(null);
/**
 * Context for dispatching events to the stream reducer.
 * Separated from state context to prevent unnecessary re-renders.
 */
const StreamDispatchContext = (0, react_1.createContext)(null);
/**
 * Context for thread management functions.
 * Separated from state to prevent unnecessary re-renders.
 */
const ThreadManagementContext = (0, react_1.createContext)(null);
/**
 * Creates initial stream state from props.
 * @param props - Provider props
 * @returns Initial stream state
 */
function createInitialState(props) {
    const { initialThread, threadId } = props;
    // Initialize with empty threadMap
    const threadMap = {};
    // If threadId is provided, initialize that thread
    if (threadId) {
        // Create initial thread state (immutably)
        const baseState = (0, event_accumulator_1.createInitialThreadState)(threadId);
        // If initial thread data provided, merge it immutably
        const threadState = initialThread
            ? {
                ...baseState,
                thread: {
                    ...baseState.thread,
                    ...initialThread,
                    id: threadId, // Always use the provided threadId
                },
            }
            : baseState;
        threadMap[threadId] = threadState;
    }
    return {
        threadMap,
        currentThreadId: threadId ?? null,
    };
}
/**
 * Provider component for stream state management.
 *
 * Uses useReducer with streamReducer to accumulate AG-UI events into
 * thread state. Provides state, dispatch, and thread management via separate contexts.
 * @returns JSX element wrapping children with stream contexts
 * @example
 * ```tsx
 * <TamboV1StreamProvider threadId="thread_123">
 *   <ChatInterface />
 * </TamboV1StreamProvider>
 * ```
 */
function TamboV1StreamProvider(props) {
    const { children } = props;
    const initialState = (0, react_1.useMemo)(() => createInitialState(props), 
    // Only recompute if threadId changes
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [props.threadId]);
    const [state, dispatch] = (0, react_1.useReducer)(event_accumulator_1.streamReducer, initialState);
    // Thread management functions
    const initThread = (0, react_1.useCallback)((threadId, initialThread) => {
        dispatch({ type: "INIT_THREAD", threadId, initialThread });
    }, []);
    const switchThread = (0, react_1.useCallback)((threadId) => {
        dispatch({ type: "SET_CURRENT_THREAD", threadId });
    }, []);
    const startNewThread = (0, react_1.useCallback)(() => {
        const tempId = `temp_${crypto.randomUUID()}`;
        // Use atomic START_NEW_THREAD action to prevent race conditions
        // when multiple calls happen concurrently (e.g., double-click)
        dispatch({ type: "START_NEW_THREAD", threadId: tempId });
        return tempId;
    }, []);
    const threadManagement = (0, react_1.useMemo)(() => ({
        initThread,
        switchThread,
        startNewThread,
    }), [initThread, switchThread, startNewThread]);
    return (react_1.default.createElement(StreamStateContext.Provider, { value: state },
        react_1.default.createElement(StreamDispatchContext.Provider, { value: dispatch },
            react_1.default.createElement(ThreadManagementContext.Provider, { value: threadManagement }, children))));
}
/**
 * Hook to access stream state.
 *
 * Must be used within TamboV1StreamProvider.
 * @returns Current stream state
 * @throws {Error} if used outside TamboV1StreamProvider
 * @example
 * ```tsx
 * function ChatMessages() {
 *   const { thread, streaming } = useStreamState();
 *
 *   return (
 *     <div>
 *       {thread.messages.map(msg => <Message key={msg.id} message={msg} />)}
 *       {streaming.status === 'streaming' && <LoadingIndicator />}
 *     </div>
 *   );
 * }
 * ```
 */
function useStreamState() {
    const context = (0, react_1.useContext)(StreamStateContext);
    if (!context) {
        throw new Error("useStreamState must be used within TamboV1StreamProvider");
    }
    return context;
}
/**
 * Hook to access stream dispatch function.
 *
 * Must be used within TamboV1StreamProvider.
 * @returns Dispatch function for sending events to reducer
 * @throws {Error} if used outside TamboV1StreamProvider
 * @example
 * ```tsx
 * function StreamHandler() {
 *   const dispatch = useStreamDispatch();
 *
 *   useEffect(() => {
 *     async function handleStream() {
 *       for await (const event of streamEvents) {
 *         dispatch({ type: 'EVENT', event });
 *       }
 *     }
 *     handleStream();
 *   }, [dispatch]);
 *
 *   return null;
 * }
 * ```
 */
function useStreamDispatch() {
    const context = (0, react_1.useContext)(StreamDispatchContext);
    if (!context) {
        throw new Error("useStreamDispatch must be used within TamboV1StreamProvider");
    }
    return context;
}
/**
 * Hook to access thread management functions.
 *
 * Must be used within TamboV1StreamProvider.
 * @returns Thread management functions
 * @throws {Error} if used outside TamboV1StreamProvider
 * @example
 * ```tsx
 * function ThreadSwitcher() {
 *   const { switchThread, startNewThread } = useThreadManagement();
 *
 *   return (
 *     <div>
 *       <button onClick={() => switchThread('thread_123')}>
 *         Load Thread
 *       </button>
 *       <button onClick={startNewThread}>
 *         New Chat
 *       </button>
 *     </div>
 *   );
 * }
 * ```
 */
function useThreadManagement() {
    const context = (0, react_1.useContext)(ThreadManagementContext);
    if (!context) {
        throw new Error("useThreadManagement must be used within TamboV1StreamProvider");
    }
    return context;
}
